core: Add C11 atomic helpers for GCC

gebart opened the pull request:
This PR adds support for GCC to use atomic operations in C11 on platforms which lack hardware support for lock-free atomics. These helper functions are required for std::atomic in C++11 on Cortex-M0, e.g. `tests/cpp11_mutex`. Without this PR we get linker errors on CM0, but not on CM3, CM4.

Tested on mulle (cortex-m4) that the functions are not linked if GCC does know how to generate lock-free atomic hardware operations.

Please test `tests/cpp11_*` on cortex-m0 platforms (nrf51822, stm32f0 etc.)

use `arm-none-eabi-objdump -d bin/stm32f0discovery/cpp11_mutex.elf | less` and search for `_atomic_` (type `/_atomic_[enter]`) to see the actual calls to the helpers.

gebart commented:
I added a compatibility commit which contains all of #3107 in order to be able to build the c++11 examples on all platforms. This committ must be removed before merge.

gebart commented:
Building application: cpp11_condition_variable (4/86) ........

		failed .. retrying: stm32f0discovery

		success: airfy-beacon, nucleo-f091, pca10000, pca10005, saml21-xpro, samr21-xpro, yunjia-nrf51822

	Building application: cpp11_mutex (5/86) ........

		failed: stm32f0discovery

		success: airfy-beacon, nucleo-f091, pca10000, pca10005, saml21-xpro, samr21-xpro, yunjia-nrf51822

	Building application: cpp11_thread (6/86) ........

		failed: stm32f0discovery

		success: airfy-beacon, nucleo-f091, pca10000, pca10005, saml21-xpro, samr21-xpro, yunjia-nrf51822

I don't understand the failure. It works on my machine for stm32f0discovery

josephnoir commented:
Which compiler version are you using and which one is used for the tests? I observed an error with gcc4.8 yesterday, but did not examine it further.

gebart commented:
```
root@792d811fb60f:/data/riotbuild# arm-none-eabi-gcc --version
arm-none-eabi-gcc (4.8.4-1+11-1) 4.8.4 20141219 (release)
Copyright (C) 2013 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

gebart commented:
@josephnoir could you paste the error message here if you can reproduce it?

josephnoir commented:
Sure, this is the `gcc version 4.8.4 20140725 (release) [ARM/embedded-4_8-branch revision 213147] (GNU Tools for ARM Embedded Processors) `from launchpad. Here is the error:
```
$ make BOARD=stm32f0discovery
ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
Warning: no PORT set!
ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
Building application "cpp11_thread" for "stm32f0discovery" with MCU "stm32f0".

ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/boards/stm32f0discovery
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/core
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/cpu/stm32f0
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/drivers
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/cpu/cortexm_common
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/cpu/stm32f0/periph
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/sys
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/sys/auto_init
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/sys/compat/hwtimer
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/sys/cpp11-compat
"/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/sys/newlib
In file included from /Users/noir/Git/RIOT/sys/include/vtimer.h:28:0,
                 from /Users/noir/Git/RIOT/sys/cpp11-compat/condition_variable.cpp:27:
/Users/noir/Git/RIOT/sys/include/timex.h: In function 'const char* timex_to_str(timex_t, char*)':
/Users/noir/Git/RIOT/sys/include/timex.h:185:39: error: 'snprintf' was not declared in this scope
              t.seconds, t.microseconds);
                                       ^
make[3]: *** [/Users/noir/Git/RIOT/tests/cpp11_thread/bin/stm32f0discovery/cpp11-compat/condition_variable.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [ALL--/Users/noir/Git/RIOT/sys/cpp11-compat] Error 2
make[1]: *** [ALL--/Users/noir/Git/RIOT/sys] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
```

gebart commented:
Seems like a missing include. Will you write a patch for timex.h or should
I?
On Jun 2, 2015 8:35 PM, "Joseph Noir" <notifications@github.com> wrote:

> Sure, this is the gcc version 4.8.4 20140725 (release)
> [ARM/embedded-4_8-branch revision 213147] (GNU Tools for ARM Embedded
> Processors)from launchpad. Here is the error:
>
> $ make BOARD=stm32f0discovery
> ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
> Warning: no PORT set!
> ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
> ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
> ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
> ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
> Building application "cpp11_thread" for "stm32f0discovery" with MCU "stm32f0".
>
> ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
> ls: /dev/tty.SLAB_USBtoUART*: No such file or directory
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/boards/stm32f0discovery
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/core
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/cpu/stm32f0
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/drivers
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/cpu/cortexm_common
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/cpu/stm32f0/periph
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/sys
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/sys/auto_init
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/sys/compat/hwtimer
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/sys/cpp11-compat
> "/Applications/Xcode.app/Contents/Developer/usr/bin/make" -C /Users/noir/Git/RIOT/sys/newlib
> In file included from /Users/noir/Git/RIOT/sys/include/vtimer.h:28:0,
>                  from /Users/noir/Git/RIOT/sys/cpp11-compat/condition_variable.cpp:27:
> /Users/noir/Git/RIOT/sys/include/timex.h: In function 'const char* timex_to_str(timex_t, char*)':
> /Users/noir/Git/RIOT/sys/include/timex.h:185:39: error: 'snprintf' was not declared in this scope
>               t.seconds, t.microseconds);
>                                        ^
> make[3]: *** [/Users/noir/Git/RIOT/tests/cpp11_thread/bin/stm32f0discovery/cpp11-compat/condition_variable.o] Error 1
> make[3]: *** Waiting for unfinished jobs....
> make[2]: *** [ALL--/Users/noir/Git/RIOT/sys/cpp11-compat] Error 2
> make[1]: *** [ALL--/Users/noir/Git/RIOT/sys] Error 2
> make[1]: *** Waiting for unfinished jobs....
> make: *** [all] Error 2
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/3115#issuecomment-108047276>.
>

josephnoir commented:
Please, go ahead! Does that mean, that the includes changed with the compiler version?

gebart commented:
rebased, testing a solution.

`#include <stdio.h>` was outside of the `extern "C"` block in timex.h. That means that the function signatures will be mangled if the header itself does not have an `extern "C"` block. Maybe there is a difference in the newlib headers between your version and mine or something similar?

gebart commented:
@josephnoir could you list your newlib and libstdc++ versions here?

```
root@b67846426411:/data/riotbuild# apt-cache policy libnewlib-arm-none-eabi
libnewlib-arm-none-eabi:
  Installed: 2.1.0+git20140818.1a8323b-2
  Candidate: 2.1.0+git20140818.1a8323b-2
  Version table:
 *** 2.1.0+git20140818.1a8323b-2 0
        500 http://httpredir.debian.org/debian/ jessie/main amd64 Packages
        100 /var/lib/dpkg/status
root@b67846426411:/data/riotbuild# apt-cache policy libstdc++-arm-none-eabi-newlib
libstdc++-arm-none-eabi-newlib:
  Installed: 4.8.3-9+4
  Candidate: 4.8.3-9+4
  Version table:
 *** 4.8.3-9+4 0
        500 http://httpredir.debian.org/debian/ jessie/main amd64 Packages
        100 /var/lib/dpkg/status
root@b67846426411:/data/riotbuild# 
```

gebart commented:
still fails on stm32f0discovery in Travis...

josephnoir commented:
I am using the version downloaded from [launchpad](https://launchpad.net/gcc-arm-embedded/4.8/4.8-2014-q3-update), which probably includes the newlib. From the metadata file, I would say it is also based on `newlib : 2.1.0 with mainline backports
    git://sourceware.org/git/newlib.git commit e14046f93c76ef701d8ad133d0ea2b96d3c1b578`. However, I did not find any info on the libstdc++ version.

josephnoir commented:
So we need a solution to solve this for Travis as it works on the boards in our tests?

gebart commented:
I want to be able to reproduce the error on my machine, but I have not had the time to install a test container with the same toolchain version as Travis.

josephnoir commented:
Ah ok, doesn't Travis use the arm-none-eabi compiler from the launchpad repository? 
```
$ sudo add-apt-repository -y ppa:terry.guo/gcc-arm-embedded
$ sudo apt-get update
$ sudo apt-get install $(./dist/tools/travis-scripts/get-pkg-list.py)
```
On their site the newest version is gcc4.9.

gebart commented:
I tested this on a new installation of Ubuntu Server 14.04.2 using the same commands as Travis to install the toolchain. The built binary inside my test VM was around 200 kBytes! When I build the same thing using my normal setup I get a binary which is around 45 kBytes. 

Digging further reveals that `__gnu_cxx::__verbose_terminate_handler` is pulled in on the "Travis" build, which in turn pulls in some tens of K of C++ name demangling and backtrace printing functions, which we can live without. I am working on a fix now.

gebart commented:
For reference: http://elegantinvention.com/blog/information/smaller-binary-size-with-c-on-baremetal-g/

gebart commented:
rebased

josephnoir commented:
Sounds good. Could you reproduce the error seen in Travis?

gebart commented:
Working on fixing it now, I have reproduced the problem on my local machine. The issue is that the code is too big to fit, and it is toolchain dependent. Using the toolchain in the riotbuild docker image does not have this issue.

gebart commented:
The problem is that the toolchain used by Travis has all of the localization extras built into its libstdc++, which means that we get a couple of KBs of text translation, numeric formats (numbers, money, dates etc.) linked into the binary. I do not know how to avoid it other than installing a different toolchain, I have not even found which configure option is responsible for the behaviour when compiling your own toolchain..

gebart commented:
I have blacklisted the boards with ROM overflows for now, the problem with the toolchain in C++ code should not affect the mergability of this C code.

gebart commented:
ping @josephnoir 

Only https://github.com/gebart/RIOT/commit/0960885b94ca3c92b8730c0eb5489006a53fc38b needs to be reviewed, the others will be deleted before merging.

gebart commented:
Using #3274 it is possible to build all C++ examples on stm32f0discovery in Travis so the previous build failures are only a problem with the installed C++ library files.

josephnoir commented:
> Only gebart@0960885 needs to be reviewed, the others will be deleted before merging.

The opening `* @{` in line 11 is never closed, the argument to `@file` in line 13 is missing and why do we need the `#if 1` in line 37?

Otherwise the code looks good.

> Using #3274 it is possible to build all C++ examples on stm32f0discovery in Travis so the previous build failures are only a problem with the installed C++ library files.

So, do want to wait for Travis to build this successfully?

Kijewski commented on core/c11_atomic.c:
```diff
@@ -0,0 +1,99 @@
+    }
+
+/* Template instantiations below */
+TEMPLATE_ATOMIC_FETCH_OP_N( add, +=, 1,  ) /* __atomic_fetch_add_1 */
```
`__atomic_fetch_add()` and friends should return the old value, shan't they? Then `op` should be `+` instead of `+=`.

gebart replied:
You are correct. I will update the PR

gebart commented:
@josephnoir 
 - `#if 1` was a leftover, removed.
 - `@file` should not have an argument, Doxygen can figure out the filename better by itself.
 - Added the missing closing Doxygen brace
 - Corrected the operators to non-updating
 - Corrected NAND to actually perform NAND and not inverted add or someting.
 - Rebased

gebart commented:
OK to remove the temporary C++ commits and squash?

josephnoir commented:
Builds for me, so go ahead.

gebart commented:
squashed, rebased, temporary commits removed.

josephnoir commented:
Jenkins won't be able to build it without the temporary C++ stuff, right?

gebart commented:
The cortex m0 platforms are already blacklisted so they won't be built, thus no errors. When this is merged we can get back to working on the C++ improvements in #3106

josephnoir commented:
The only Travis failure is on the m0 builds, where nothing happens. So this should be fine.

gebart commented:
OK, Travis is green now.

@josephnoir Is that an ACK?

This needs two ACKs because this is a core change. @kaspar030 @Kijewski @OlegHahm ?

kaspar030 commented:
Looks good to me. ACK.

josephnoir commented:
Yes, ACK.
